universal/include/userver/formats/universal/common_containers.hpp

```diff
@@ -0,0 +1,296 @@
+struct Default : public impl::EmptyCheck, public impl::Param<T> {
+  inline constexpr Default(const T& value) :
+      impl::Param<T>(value) {}
+};
+template <utils::ConstexprString Pattern>
+static const utils::regex kRegex(Pattern);
+
+struct Pattern : public impl::EmptyCheck, public impl::Param<const utils::regex*> {
+  constexpr inline Pattern(const utils::regex& regex) :
+      impl::Param<const utils::regex*>(&regex) {}
+  constexpr inline std::string Check(std::string_view str) const {
+    return utils::regex_match(str, *this->value) ? "" : this->Error(str);
+  }
+  constexpr inline std::string Error(std::string_view) const {
+    return "Error";
+  }
+};
+struct Additional : public impl::EmptyCheck, public impl::Param<bool> {
+  constexpr inline Additional(const bool& value) :
+      impl::Param<bool>(value) {}
+};
+template <>
+struct FieldConfig<int> {
+  std::optional<Max<int>> Maximum;
+  std::optional<Min<int>> Minimum;
+  template <typename MainClass, auto I, typename Value>
+  constexpr int Read(Value&& value) const {
+    constexpr auto name = boost::pfr::get_name<I, MainClass>();
+    return value[name].template As<int>();
+  };
+  template <typename MainClass, auto I, typename Value>
+  constexpr std::optional<int> TryRead(Value&& value) const {
+    constexpr auto name = boost::pfr::get_name<I, MainClass>();
+    return parse::TryParse(value[name], parse::To<int>{});
+  };
+  constexpr auto Write(const int& value, std::string_view fieldName, const auto&, auto& builder) const {
+    builder[static_cast<std::string>(fieldName)] = value;
+  };
+  inline constexpr std::string_view Check(const int&) const {
+    return "";
+  }
+
+};
+template <>
+struct FieldConfig<std::optional<std::string>> {
+  std::optional<Pattern> Pattern;
+  std::optional<Default<std::string>> Default;
+  template <typename MainClass, auto I, typename Value>
+  constexpr std::optional<std::string> Read(Value&& value) const {
+    constexpr auto name = boost::pfr::get_name<I, MainClass>();
+    if(!value.HasMember(name)) {
+      return std::nullopt;
+    };
+    return value[name].template As<std::string>();
+  };
+  template <typename MainClass, auto I, typename Value>
+  constexpr std::optional<std::optional<std::string>> TryRead(Value&& value) const {
+    constexpr auto name = boost::pfr::get_name<I, MainClass>();
+    auto response = parse::TryParse(value[name], parse::To<std::string>{});
+    if(response) {
+      return response;
+    }
+    if(this->Default) {
+      return this->Default->value;
+    }
+    return std::nullopt;
+  }
+  constexpr auto Write(const std::optional<std::string>& value, std::string_view fieldName, const auto&, auto& builder) const {
+    if(value) {
+      builder[static_cast<std::string>(fieldName)] = *value;
+    };
+  };
+  inline constexpr std::string_view Check(const std::string&) const {
+    return "";
+  }
+
+};
+template <>
+struct FieldConfig<std::string> {
+  std::optional<Pattern> Pattern;
```

Comment:
Non-`std::optional` specializations could make use of `Default` during parsing as well. Then through some usage of `std::conditional_t` and `if constexpr`, the `optional` and non-`optional` specializations can be merged

Reply:
Actually, no, see the comment above that talks about null and missing in `optional`

`optional` definitely needs some extra properties, at least for nullability. Let's keep duplicate `FieldConfig`s for optional and non-optional types for now. We'll return to this question later

Reply:
After a third thought, no, let's not add `Default` to non-`optional` fields